modepy/shapes.py

```diff
@@ -559,4 +566,40 @@ def _submes_for_hypercube(shape: Hypercube, node_tuples):
 # }}}
 
 
+# {{{ tensor product shape
+
+@dataclass(frozen=True, init=False)
+class TensorProductShape(Shape):
+    """
+    .. attribute:: bases
+
+        A :class:`tuple` of base shapes that form the tensor product.
+    """
+
+    def __init__(self, bases: Tuple[Shape, ...]) -> None:
+        self.bases = bases
+
+    @property
+    def dim(self) -> int:
+        return sum(s.dim for s in self.bases)
+
+    @property
+    def nvertices(self) -> int:
+        return np.prod([s.nvertices for s in self.bases])
+
+    @property
+    def nfaces(self) -> int:
+        # FIXME: is there a formula for this?
```

Comment:
For "adventurous shape" x interval, seemingly yes: One per face of the adventurous shape, plus a lid on each end. I believe this can be repeated for any number of intervals tensor-producted onto the shape.

Reply:
Yep, implemented that in b7db9ff.